src/server/list_family.cc

```diff
@@ -241,90 +145,30 @@ class BPopPusher {
   ListDir popdir_, pushdir_;
 };
 
-BPopper::BPopper(ListDir dir) : dir_(dir) {
-}
-
-OpStatus BPopper::Run(Transaction* trans, unsigned msec) {
-  auto tp = msec ? chrono::steady_clock::now() + chrono::milliseconds(msec) : time_point::max();
-  bool is_multi = trans->IsMulti();
-
-  trans->Schedule();
-
-  auto* stats = ServerState::tl_connection_stats();
-
-  OpResult<ShardFFResult> result = FindFirst(trans);
-
-  if (result.ok()) {
-    ff_result_ = move(result.value());
-  } else if (result.status() == OpStatus::KEY_NOTFOUND) {
-    // Close transaction and return.
-    if (is_multi) {
-      auto cb = [](Transaction* t, EngineShard* shard) { return OpStatus::OK; };
-      trans->Execute(std::move(cb), true);
-      return OpStatus::TIMED_OUT;
-    }
-
-    auto wcb = [](Transaction* t, EngineShard* shard) {
-      return t->GetShardArgs(shard->shard_id());
-    };
-
-    VLOG(1) << "Blocking BLPOP " << trans->DebugId();
-    ++stats->num_blocked_clients;
-    bool wait_succeeded = trans->WaitOnWatch(tp, std::move(wcb));
-    --stats->num_blocked_clients;
-
-    if (!wait_succeeded)
-      return OpStatus::TIMED_OUT;
-  } else {
-    // Could be the wrong-type error.
-    // cleanups, locks removal etc.
-    auto cb = [](Transaction* t, EngineShard* shard) { return OpStatus::OK; };
-    trans->Execute(std::move(cb), true);
-
-    DCHECK_NE(result.status(), OpStatus::KEY_NOTFOUND);
-    return result.status();
-  }
-
-  auto cb = [this](Transaction* t, EngineShard* shard) {
-    Pop(t, shard);
-    return OpStatus::OK;
-  };
-  trans->Execute(std::move(cb), true);
-
-  return OpStatus::OK;
-}
-
-void BPopper::Pop(Transaction* t, EngineShard* shard) {
-  if (auto wake_key = t->GetWakeKey(shard->shard_id()); wake_key) {
-    key_ = *wake_key;
-    OpPop(t, shard);
-  } else if (shard->shard_id() == ff_result_.sid) {
-    ff_result_.key.GetString(&key_);
-    OpPop(t, shard);
-  }
-}
-
-void BPopper::OpPop(Transaction* t, EngineShard* shard) {
+// Called as a callback from MKBlocking after we've determined which key to pop.
+std::string OpBPop(Transaction* t, EngineShard* shard, std::string_view key, ListDir dir) {
```

Comment:
yeah :see_no_evil: 

It's what I'm used to, but it doesn't really matter for types. For function calls I do prefer qualifying to avoid ADL bugs and (but no idea if it's even measurable) to reduce the lookup work the compiler has to do.